cmd/hub/mcscontroller/controller_test.go

```diff
@@ -0,0 +1,158 @@
+package mcscontroller_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/submariner-io/lighthouse/cmd/hub/mcscontroller"
+	lhconst "github.com/submariner-io/lighthouse/pkg/constants"
+	"github.com/submariner-io/lighthouse/pkg/lhutil"
+	"github.com/submariner-io/lighthouse/pkg/mcs"
+)
+
+const (
+	serviceName = "svc"
+	serviceNS   = "svc-ns"
+	brokerNS    = "submariner-broker"
+	cluster1    = "c1"
+	cluster2    = "c2"
+)
+
+var (
+	service = &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: serviceNS,
+		},
+	}
+	export = &mcsv1a1.ServiceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "placeholder",
+			Namespace: brokerNS,
+			Labels: map[string]string{
+				lhconst.LabelSourceNamespace:      serviceNS,
+				lhconst.LighthouseLabelSourceName: serviceName,
+			},
+		},
+	}
+)
+
+func TestImportGenerated(t *testing.T) {
+	// @todo refactor into a prepareServiceExport function
+	exp1 := export.DeepCopy()
+	exp1.Name = lhutil.GenerateObjectName(serviceName, serviceNS, cluster1)
+	exp1.Labels[lhconst.LighthouseLabelSourceCluster] = cluster1
+	es, _ := mcs.NewExportSpec(service, exp1, cluster1)
+	_ = es.MarshalObjectMeta(&exp1.ObjectMeta)
+
+	preloadedObjects := []runtime.Object{service, exp1}
+	ser := mcscontroller.ServiceExportReconciler{
+		Client: getClient(preloadedObjects),
+		Log:    newLogger(t, false),
+		Scheme: getScheme(),
+	}
+
+	req := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      exp1.GetName(),
+			Namespace: exp1.GetNamespace(),
+		}}
+
+	result, err := ser.Reconcile(context.TODO(), req)
+
+	if err != nil {
+		t.Error(err)
+	}
+	if result.Requeue {
+		t.Error("unexpected requeue")
+	}
+
+	si := mcsv1a1.ServiceImport{}
+	err = ser.Client.Get(context.TODO(), req.NamespacedName, &si)
+
+	if err != nil {
+		t.Error(err)
+	}
+	// @todo validate the ServiceImport properties
+}
+
+// return a scheme with the default k8s and mcs objects
+func getScheme() *runtime.Scheme {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(mcsv1a1.AddToScheme(scheme))
+	return scheme
+}
+
+// generate a fake client with preloaded objects
+func getClient(objs []runtime.Object) client.Client {
+	return fake.NewClientBuilder().WithScheme(getScheme()).WithRuntimeObjects(objs...).Build()
+}
+
+// satisfy the logr.Logger interface with a nil logger
```

Comment:
why a nil logger and not using an actual logger? those messages can be important for debugging..
should you really implement the whole logger interface?

Reply:
changed in subsequent commit to log via `t.Error` and `t.Log`

Reply:
the code expects the full interface so unless I find a logger I can embed, needed to reimplement the interface